cgroup v2 docs (#172)
kinvolk/flatcar-docs: docs/container-runtimes/switching-to-unified-cgroups.md

---
title: Switching to Unified Cgroups
linktitle: Switching to unified cgroups
description: Overview of changes necessary to use unified cgroups with Kubernetes
weight: 20
aliases:
---

> With the upgrade to Systemd v248, Flatcar Linux has migrated to the unified

Reviewer: We might not migrate to systemd v248, rather we would directly jump to v249. Do you think if it will confuse people if we mentioned that since `systemd v248, Flatcar Linux has migrated to the unified cgroup hierachy`?

Author: I didn't know how to express this better - do you think i should just put a provisional release number in (2970?) and then correct if the release happens a day +/-?

Reviewer: Yes, I think using a provisional release number would be better compared to systemd version.

Author: Done

cgroup hierarchy (aka cgroup v2). Much of the container ecosystem has already
moved to default to cgroup v2. Cgroup v2 brings new and exciting features in
areas such as eBPF and rootless containers.

Flatcar nodes deployed prior to this change will be kept on cgroup v1 (legacy
hierarchy) and will require manual migration. During an update from an older
Flatcar version, a post update script does two things:

* adds the kernel command line parameters `systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller`
  to `/usr/share/oem/grub.cfg`
* creates a systemd drop-in unit at `/etc/systemd/system/containerd.service.d/10-use-cgroupfs.conf` that

>   configures `containerd` to use systemd for cgroups.

Reviewer 2: ```suggestion
  configures `containerd` to use `cgroupfs` for cgroups.
```

Author: Thanks, found that one just a second ago.

# Migrating old nodes to unified cgroups

To undo the changes performed by the post update script, execute the following commands as root (or using `sudo`):

```bash
rm /etc/system/system/containerd.service.d/10-use-cgroupfs.conf

> sed -i -e '/systemd.unified_cgroup_hierarchy=0/d' /usr/share/oem/grub.cfg

Reviewer 2: ```suggestion
sed -i -e '/systemd.unified_cgroup_hierarchy=0/d' /usr/share/oem/grub.cfg
sed -i -e '/systemd.legacy_systemd_cgroup_controller/d' /usr/share/oem/grub.cfg
```

```json
{
  "exec-opts": ["native.cgroupdriver=systemd"]
}
```
or adding a `docker.service` drop-in at `/etc/systemd/system/docker.service.d/10-cgroup-v2.conf`:
```ini
[Service]
Environment="DOCKER_CGROUPS=--exec-opt native.cgroupdriver=systemd"
```

## Container Runtimes

When deploying Kubernetes through `kubeadm`, the default container runtime on Flatcar will be `dockershim`. In this setup, `kubelet` talks to `dockershim`, which talks to `docker`, which interfaces with `containerd`. The `SystemdCgroup` setting in `containerd`'s `config.toml` is ignored. `docker`'s cgroup driver and `kubelet` cgroup driver settings must match. Starting with Kubernetes v1.22, `kubeadm` will default to the `systemd` `cgroupDriver` setting if no setting is provided explicitly. Out of the box, Flatcar defaults are compatible with Docker and Kubernetes defaults - everything will use `systemd` cgroup driver.

When using `kubeadm`, add the snippet to your `kubeadm-config.yaml` to configure the `kubelet` cgroup driver:

```yaml
---
kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: systemd
```

## Containerd

> If users choose the `containerd` runtime, they must ensure that `containerd`'s setting for `SystemdCgroup` is consistent with `kubelet` and `docker` settings. Flatcar enables `SystemdCgroup` by default for `containerd`. Users may change the setting to suit their deployment. To do that, follow the instructions on
> [how to customize containerd configuration](customizing-docker) and add the relevant lines to your `config.toml`:

Reviewer 2: ```suggestion
If users choose the `containerd` runtime, they must ensure that `containerd`'s setting for `SystemdCgroup` is consistent with `kubelet` and `docker` settings. Flatcar enables `SystemdCgroup` by default for `containerd`. Users may change the setting to suit their deployment.
If you maintain your own containerd configuration or did follow the instructions on
[how to customize containerd configuration](customizing-docker), you should add the relevant lines to your `config.toml`:
```

---
title: Switching to Unified Cgroups
linktitle: Switching to unified cgroups
description: Overview of changes necessary to use unified cgroups with Kubernetes
weight: 20
aliases:
---

Beginning with Flatcar version 2970.0.0, Flatcar Linux has migrated to the unified
cgroup hierarchy (aka cgroup v2). Much of the container ecosystem has already
moved to default to cgroup v2. Cgroup v2 brings exciting new features in
areas such as eBPF and rootless containers.

Flatcar nodes deployed prior to this change will be kept on cgroups v1 (legacy
hierarchy) and will require manual migration. During an update from an older
Flatcar version, a post update script does two things:

* adds the kernel command line parameters `systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller`
  to `/usr/share/oem/grub.cfg`
* creates a systemd drop-in unit at `/etc/systemd/system/containerd.service.d/10-use-cgroupfs.conf` that
  configures `containerd` to keep using cgroupfs for cgroups.

# Migrating old nodes to unified cgroups

To undo the changes performed by the post update script, execute the following commands as root (or using `sudo`):

```bash

> rm /etc/system/system/containerd.service.d/10-use-cgroupfs.conf

Author: ```suggestion
rm /etc/systemd/system/containerd.service.d/10-use-cgroupfs.conf
```

---
title: Switching to Unified Cgroups
linktitle: Switching to unified cgroups
description: Overview of changes necessary to use unified cgroups with Kubernetes
weight: 20
aliases:
---

With the upgrade to Systemd v248, Flatcar Linux has migrated to the unified
cgroup hierarchy (aka cgroup v2). Much of the container ecosystem has already
moved to default to cgroup v2. Cgroup v2 also brings new and exciting features related to
eBPF tracing.

Flatcar nodes deployed prior to this change will be kept on cgroup v1 (legacy
hierarchy) and will require manual migration. During an update from an older
Flatcar version, a post installation script adds the kernel command line
parameter `systemd.unified_cgroup_hierarchy=0` to `/usr/share/oem/grub.cfg`. To
migrate such nodes to cgroup v2, either remove the line or change it to
`systemd.unified_cgroup_hierarchy=1`.

Newly deployed nodes will default to cgroup v2.  To revert to cgroup v1 on such

> nodes, use the following ignition snippet:

Reviewer 2: ```suggestion
nodes, use the following Ignition snippet (here as CLC YAML to be transpiled to Ignition JSON):
```

```yaml
storage:
  filesystems:
    - name: "OEM"
      mount:
        device: "/dev/disk/by-label/OEM"
        format: "btrfs"
  files:
    - filesystem: "OEM"
      path: "/grub.cfg"
      mode: 0644
      append: true
      contents:
        inline: |
          set linux_append="$linux_append systemd.unified_cgroup_hierarchy=0"
```

> A reboot is required before the snippet becomes active.

Reviewer 2: ```suggestion
However, this setting doesn't take effect on the first boot, and a reboot is required before the snippet becomes active.
```

```ini
[Service]
Environment="DOCKER_CGROUPS=--exec-opt native.cgroupdriver=systemd"
```

# Kubernetes Container Runtimes

When deploying Kubernetes through `kubeadm`, the default container runtime on Flatcar will be `dockershim`. In this setup, `kubelet` talks to `dockershim`, which talks to `docker`, which interfaces with `containerd`. The `SystemdCgroup` setting in `containerd`'s `config.toml` is ignored. `docker`'s cgroup driver and `kubelet` cgroup driver settings must match. Starting with Kubernetes v1.22, `kubeadm` will default to the `systemd` `cgroupDriver` setting if no setting is provided explicitly. Out of the box, Flatcar defaults are compatible with Docker and Kubernetes defaults - everything will use `systemd` cgroup driver.

If users choose the `containerd` runtime, they must ensure that `containerd`'s support for `SystemdCgroups` is enabled. To do that, follow the instructions on
[how to customize containerd configuration](customizing-docker) and add the following to `config.toml`:
```toml
[plugins.cri.containerd.runtimes.runc]
  [plugins.cri.containerd.runtimes.runc.options]

>     SystemdCgroup = true

Reviewer 2: Can we add this as default in Flatcar's containerd config?

Author: I thought about it - but changing this breaks kubernetes clusters because a cgroupfs-kubelet will not tolerate this value being SystemdCgroup. The one thing we could do is make this work like `lokomotive` here https://github.com/kinvolk/lokomotive/blob/master/assets/terraform-modules/node/templates/node.yaml.tmpl#L118 - have a script run before containerd starts that generates an config snippet that is included. What do you think?

Reviewer 2: If I understand it correctly, the default is cgroupfs and won't work for new deployments? That's not really ideal… Yes, writing out the right choice for the runtime-generated containerd config file is an option.

The other is that we can add an entry here, too, but still people would need to find it:
https://kinvolk.io/docs/flatcar-container-linux/latest/container-runtimes/switching-from-docker-to-containerd-for-kubernetes/

Reviewer 2: Maybe I didn't get the part of the default value, in case containerd selects systemd as default after a cgroupv2 environment was found, then we should be fine without explicitly adding an entry, but if that it's the case we would need to have a script for that in torcx.

Reviewer 2: Ok, I think this here is fine since there is no autodiscovery but we need to call it out in https://kinvolk.io/docs/flatcar-container-linux/latest/container-runtimes/switching-from-docker-to-containerd-for-kubernetes/, too.

I would also like to have a date where the default gets switched to `true`, and have this stated in the beginning of this doc (and possibly in the motd).

Author: When docker is the CRI, the containerd SystemdCgroup setting has no effect. Docker does autodetection, containerd doesn't.  I was originally thinking we can get away with keeping the containerd setting in at the default, and changing it 2 releases (1 month) later, but now i'm not so sure. I fear that adding autodetection increases the complexity. I need to sleep on it to have a better idea.
For sure i'll update the docs you referenced.

Reviewer 2: Yes, autodetection is not the best idea, and it's also not easy because it has to consider whether cgroupv1 or cgroupv2 is used and if cgroupv1 is used, whether the kubelet wants to use `cgroupfs` or `systemd`…

As said, the compromise I see is to keep the default to `cgroupfs` to not break existing systems but also communicate that the default will be switched.
Another option is yet antoher postinstall hook where we could freeze the containerd config (in case people use the one from torcx) to `cgroupfs` and do the switch to systemd now.

Reviewer 2: The decision was to extend the postinstall hook to also freeze the containerd config: https://github.com/kinvolk/update_engine/pull/13
